server/jetstream_cluster.go

```diff
@@ -1806,6 +1830,48 @@ func (js *jetStream) monitorStream(mset *stream, sa *streamAssignment, sendSnaps
 					stopMigrationMonitoring()
 				}
 			}
+
+			// Here we are checking if we are not the leader but we have been asked to allow
+			// direct access. We now allow non-leaders to participate in the queue group.
+			if !isLeader && mset != nil {
+				mset.mu.Lock()
+				// Check direct gets first.
+				if mset.cfg.AllowDirect {
+					if mset.directSub == nil && mset.isCurrent() {
+						mset.subscribeToDirect()
+					} else {
+						startDirectAccessMonitoring()
+					}
+				}
+				// Now check for mirror directs as well.
+				if mset.cfg.MirrorDirect {
+					if mset.mirror != nil && mset.mirror.dsub == nil && mset.isCurrent() {
+						mset.subscribeToMirrorDirect()
+					} else {
+						startDirectAccessMonitoring()
+					}
+				}
+				mset.mu.Unlock()
+			}
+
+		case <-datc:
+			mset.mu.Lock()
+			ad, md, current := mset.cfg.AllowDirect, mset.cfg.MirrorDirect, mset.isCurrent()
+			if !current {
+				mset.mu.Unlock()
+				continue
+			}
+			// We are current, cancel monitoring and create the direct subs as needed.
+			if ad {
+				mset.subscribeToDirect()
+			}
+			if md {
+				mset.subscribeToMirrorDirect()
+			}
+			mset.mu.Unlock()
+			// Stop monitoring.
+			stopDirectMonitoring()
```

Comment:
So this monitoring is good but if I read it right its only activated on leader change, stops once caught up and then never checked again?  What if a node lags later? Likewise later where you check based on dgetCaughtUpThresh.

Really concerned about read after write consistency here for KV - so this wouold have to be a opt-in for specific KV cases where thats acceptible? When I designed KV you were pretty much saying read after write consistency is non negotiable - even the cache design being pass through not write through had to maintain that.

Reply:
I have been thinking about that and whether or not we want to do anything there. We do not do anything like that for normal consumers, at least so far. If the server's connection is broke to a route etc the subscription will be dropped, meaning this subscriber will not participate while disconnected.

This is opt-in yes, and read after write could be an issue, and yes I was the one arguing about it for sure. I wonder if the use case is simply not common. I was using gmail as a use case, but they do not even do that now and short circuit. Meaning is you were keeping a list of keys and you updated one, either you would hand update the list, or update it once a watcher fired. Not sure you would do put then get right behind it. WDYT?

Reply:
Agree read-after-write is not as important as initally imagined, however its out there now and cannot change by default.  

At best we can have an opt in to relax that constraint for the more kv abused as large pub-sub with cache behavior people are using it for thats causing the problems.

Reply:
I was leaning for making this default and you could op-out. With this approach you can just add mirrors and things get better, if its an opt-in all consumers would need to be updated before the mirrors would be used for get operations.

Reply:
Read-after-write aside, we definitely need to know its not using a !current mirror, definitely think that requires some more thought - mirrors especially can silently and forever stop working should the source be replaced (which. makes sense of course) but without that scenario being detected and that mirror declaring itself not a read candidate its trouble.

Same scenario with excessive lag or just long term badness